pkg/plugin/resource_handler.go

```diff
@@ -0,0 +1,212 @@
+	mux.HandleFunc("/resourcegroups", ocidx.GetResourceGroupHandler)
+	mux.HandleFunc("/dimensions", ocidx.GetDimensionsHandler)
+	mux.HandleFunc("/tags", ocidx.GetTagsHandler)
+}
+
+func (ocidx *OCIDatasource) GetTenanciesHandler(rw http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodGet {
+		respondWithError(rw, http.StatusMethodNotAllowed, "Invalid method", nil)
+		return
+	}
+
+	// ts := ocidx.clients.GetTenancies(req.Context())
+	ts := ocidx.GetTenancies(req.Context())
+
+	writeResponse(rw, ts)
+}
+
+func (ocidx *OCIDatasource) GetRegionsHandler(rw http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodPost {
+		respondWithError(rw, http.StatusMethodNotAllowed, "Invalid method", nil)
+		return
+	}
+
+	var rr rootRequest
+	if err := jsoniter.NewDecoder(req.Body).Decode(&rr); err != nil {
+		backend.Logger.Error("plugin.resource_handler", "GetRegionsHandler", err)
+		respondWithError(rw, http.StatusBadRequest, "Failed to read request body", err)
+		return
+	}
+	regions := ocidx.GetSubscribedRegions(req.Context(), rr.Tenancy)
+
+	writeResponse(rw, regions)
+}
+
+func (ocidx *OCIDatasource) GetCompartmentsHandler(rw http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodPost {
+		respondWithError(rw, http.StatusMethodNotAllowed, "Invalid method", nil)
+		return
+	}
+
+	var rr rootRequest
+	if err := jsoniter.NewDecoder(req.Body).Decode(&rr); err != nil {
+		backend.Logger.Error("plugin.resource_handler", "GetCompartmentsHandler", err)
+		respondWithError(rw, http.StatusBadRequest, "Failed to read request body", err)
+		return
+	}
+	compartments := ocidx.GetCompartments(req.Context(), rr.Tenancy)
+
+	writeResponse(rw, compartments)
+}
+
+func (ocidx *OCIDatasource) GetNamespacesHandler(rw http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodPost {
+		respondWithError(rw, http.StatusMethodNotAllowed, "Invalid method", nil)
+		return
+	}
+
+	var nmr namespaceMetricRequest
+	if err := jsoniter.NewDecoder(req.Body).Decode(&nmr); err != nil {
+		backend.Logger.Error("plugin.resource_handler", "GetNamespacesHandler", err)
+		respondWithError(rw, http.StatusBadRequest, "Failed to read request body", err)
+		return
+	}
+
+	namespaces := ocidx.GetNamespaceWithMetricNames(req.Context(), nmr.Tenancy, nmr.Compartment, nmr.Region)
+
+	writeResponse(rw, namespaces)
+}
+
+func (ocidx *OCIDatasource) GetResourceGroupHandler(rw http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodPost {
+		respondWithError(rw, http.StatusMethodNotAllowed, "Invalid method", nil)
+		return
+	}
+
+	var rgr resourceGroupRequest
+	if err := jsoniter.NewDecoder(req.Body).Decode(&rgr); err != nil {
+		backend.Logger.Error("plugin.resource_handler", "GetResourceGroupHandler", err)
+		respondWithError(rw, http.StatusBadRequest, "Failed to read request body", err)
+		return
+	}
+
+	rgs := ocidx.GetResourceGroups(req.Context(), rgr.Tenancy, rgr.Compartment, rgr.Region, rgr.Namespace)
```

Comment:
If GetResourceGroups() returns nil ?

Reply:
It is fine, ResourceGroups is not mandatory, and in many case there aren't resource groups in metrics.